include/pybind11/stl_bind.h

```diff
@@ -595,13 +595,33 @@ template <typename Map, typename Class_> auto map_if_insertion_operator(Class_ &
     );
 }
 
+template<typename Map>
```

Comment:
It appears these types are used exclusively inside bind_map. 
Does it work if we define these types inside bind_map, e.g.

replace the `using` line with
```
struct KeysView {
  Map &map;
};
```

Reply:
Possibly - what is the advantage though? The disadvantage is that the class will probably get a much less readable name in compiler errors.

Reply:
The motivation would be to minimize the scope of name declarations.

If we have to use the details namespace as a workaround for compiler not reporting readable errors, then let's keep it as is ... :)

Reply:
> The motivation would be to minimize the scope of name declarations.

That makes sense. I'll give it a try tomorrow.

Reply:
MSVC did not like it: lots of warnings like this, which made the build fail.
```
D:\a\pybind11\pybind11\include\pybind11/pybind11.h(1620): warning C4503: 'std::_Compressed_pair<_Dx,pybind11::bind_map::items_view *,true>': decorated name length exceeded, name was truncated [D:\a\pybind11\pybind11\build\tests\pybind11_cross_module_tests.vcxproj]
```

Reply:
Ah. MSVC strikes again! The lesson suggests we shall minimize the use of inlined / nested types. 

Thanks for trying.